### ISSUE TITLE
QuickFormat reference panel positioning on windows

### ISSUE THREAD

**Issue by abaevbog** [zotero/zotero#3847]
@adomasven, I'm afraid I just ran into the exact reason for the small change to the margins that [broke the dialog](https://github.com/zotero/zotero/commit/633bee45afce3a4f68476ccc205cdc607e126035#commitcomment-139546386) on windows with non-standard refresh rate.

This is the reference panel positioning issue I'm seeing on windows 11 VM:

<img width="896" alt="Screenshot 2024-03-14 at 12 40 58 AM" src="https://github.com/zotero/zotero/assets/36271954/7d1ad653-bdfa-4573-8f05-5a3381ee2044">

Can you reproduce it? This does not happen for me with `dialog.style.marginBottom = "-36px";` that caused refresh rate breakage, so fixing this is what must have been my intention then. 

Now that there's more context, I'll try different options of panel positioning that hopefully can fix it without breaking things elsewhere

**Comment by abaevbog** [comment]
I tried [this tweak](https://github.com/abaevbog/zotero/commit/5fe9c8a24a387c7d89b86c7a840f12d3cb99ff3d) to use the editor, as opposed to `document.documentElement`, as the anchor and it did position the panel correctly. This still requires a few more tweaks to make sure that as the window expands/shrinks the panel is properly placed (especially if the ref panel is always open). 

I didn't want to go too far without making sure that it does not bring aback the refresh rate breakage, and I still have not been able to fully reproduce that on my monitor. So I wanted to check in if the dialog properly appears with this change for you?

**Comment by adomasven** [comment]
> I didn't want to go too far without making sure that it does not bring aback the refresh rate breakage, and I still have not been able to fully reproduce that on my monitor. So I wanted to check in if the dialog properly appears with this change for you?

I was only able to reproduce it on Windows 10 running natively with an nVidia GPU. I do not have a setup with a VM that could run at higher refresh rates. I'll test this and let you know.

**Comment by adomasven** [comment]
It opens fine with the tweak on my setup.

**Comment by abaevbog** [comment]
Ok, awesome! As far as I can tell, mac and linux work work as well. I added a small tweak to make sure that after window resizing the panel is closed and re-opened so that it gets repositioned properly.

I'm a bit surprised that we didn't see any reports about this, since it makes all lines of the editor but the first one very hard to use... I guess not that many people end up having so many bubbles that they need more than a single line in the editor.

**Comment by abaevbog** [comment]
This is probably a better place to bring up that the high refresh rate is still causing problems sometimes: https://forums.zotero.org/discussion/comment/459517/#Comment_459517.

I wanted to share a couple of ideas that we could try:
- we could go back to setting the margin top on the reference panel and just dynamically increase/decrease it so that the panel is somewhat reasonably positioned. Something like this https://github.com/abaevbog/zotero/commit/60c7c7e6d2f067687d6e15ce46da6dd3ef48ae38
- if the issue has something to do with the nature of the `<panel>`, we could make it a simple `<vbox>` and just display/hide/resize it as needed. I have a general proof of concept working here https://github.com/abaevbog/zotero/commit/eef9654361444f7347be90e94d53069a134e9d31. I sorta like this one because it also allows us to not worry about `noautohide`, `noautofocus` and it guarantees that the references list won't break away from the main dialog. It is a bit more work though.
- finally, I suspect that these issues may have something to do with `drawintitlebar`/`hidechrome`. At least I believe this is what makes us add all these margins in the first place. The title bar appears on linux, so I'm wondering if there's a reason why they must be hidden on mac and win?

**Comment by adomasven** [comment]
Unfortunately I'm unable to reproduce this.

> * finally, I suspect that these issues may have something to do with `drawintitlebar`/`hidechrome`. At least I believe this is what makes us add all these margins in the first place. The title bar appears on linux, so I'm wondering if there's a reason why they must be hidden on mac and win?

We will definitely get rid of the red border once we redesign the new citation dialog, and I think we can get rid of the red border for Zotero 7, especially since it's causing all sorts of problems. I know it's somewhat iconic, but also quite archaic, and we may as well remove it now with good reason.

**Comment by adomasven** [comment]
Ok, we may not want to remove the red border, in that case

> if the issue has something to do with the nature of the <panel>, we could make it a simple <vbox> and just display/hide/resize it as needed. I have a general proof of concept working here https://github.com/abaevbog/zotero/commit/eef9654361444f7347be90e94d53069a134e9d31. I sorta like this one because it also allows us to not worry about noautohide, noautofocus and it guarantees that the references list won't break away from the main dialog. It is a bit more work though.

Would this make it so that #3889 is required or simper? If so, I think we should go for that. There's still a caveat of not knowing whether it will not break for some people, but if it's less hacky and using negative margins, it's probably more likely to work, and we will still eventually need this to work somehow for the redesign.

**Comment by windingwind** [comment]
Any reason we use the `WindowDraggingElement` instead of the `-moz-window-dragging` and manually resize the window? These designs contribute to most of the positioning problems.

**Comment by abaevbog** [comment]
@windingwind so the `WindowDraggingElement` has been around from way before the bulk for recent updates (most started as accessibility-focused tweaks but it sorta evolved into a much bigger rewrite), and it just didn't get in the way of those changes. I did notice during testing that dragging was glitchy but I blamed it on my windows VM. If there is a way no not use it, that would be great I think. I tried to set `-moz-window-dragging: drag` on the main top-most `<box>` and the dragging works for me on mac but not on windows. Am I missing something else with it?

In regards to the manual resizing, are you referring to [this](https://github.com/zotero/zotero/blob/45bd3745301115286c528858f60268183635a068/chrome/content/zotero/integration/quickFormat.js#L1213)? The editor with bubbles can expand or shrink depending on how much content there is in it. The window has to be resized, otherwise the editor can go from, say, 60px height to 90px height when you type something, and if the window does not change, the editor will get cut-off. The same is true in the opposite direction: if you have a tall editor with several line of bubbles and you delete all of them, the editor and the entire dialog needs to shrink to not have the empty space.

**Comment by windingwind** [comment]
The `-moz-window-dragging: drag` works for me on Windows (with the `WindowDraggingElement` disabled and `chromemargin="0,2,2,2"`, which is platform-specific, see our main window). However, the resize thing cuts off the window after dragging.

So what I have in mind is, using a `<panel>` to replace the actual window, where it gets automatically resized according to the content size, the same as the non-native menu popup, the add item by identifier, and the opened tabs panel. Haven't tried but hopefully, that can get rid of the resizing and all its related issues.

**Comment by abaevbog** [comment]
Hmm, not sure what's up with the dragging. Maybe it is the VM cause on my mac it works just fine. I'll try to trouble shoot it later. 

> However, the resize thing cuts off the window after dragging. 

That part is strange because the window resizing is only supposed to happen if something within the editor changes (e.g. bubbled added/removed/etc.). There is [one call ](https://github.com/zotero/zotero/blob/45bd3745301115286c528858f60268183635a068/chrome/content/zotero/integration/quickFormat.js#L1996) to resize the window after a bubble is stopped being dragged - but that should not trigger if the actual window is being dragged

**Comment by abaevbog** [comment]
> Would this make it so that https://github.com/zotero/zotero/pull/3889 is required or simper? If so, I think we should go for that. There's still a caveat of not knowing whether it will not break for some people, but if it's less hacky and using negative margins, it's probably more likely to work, and we will still eventually need this to work somehow for the redesign.

@adomasven making the reference panel a `vbox` instead of a `panel` would indeed replace most of my current changes for `fx115` - apart from a few css tweaks. That said, most of those changed were related to the dragging logic to deal with the reference panel not following the main window. So if `-moz-window-dragging: drag` works better as an alternative to current `WindowDraggingElement` and none of those changes are needed - all `fx115` would require is some tweaks to width css properties.

But in any case, since we don't actually know the root of the refresh-rate issue (and nobody seems to be able to reliably reproduce it), it still might be worth a shot?
I think that commit is fairly close to being done, so I'll try to wrap it up and we can see if it makes any kinda improvement.

**Comment by abaevbog** [comment]
> Does replacing the <panel> element with a <vbox> still allow it to float outside the bounds of the actual window

No, it's all just one window that includes both the editor and the references list. Just now instead of resizing the window only due to changes in the editor, I also do some resizing to keep the window in sync with the height of the panel. So, after a query, we see that the references list + the editor is too tall / too short for this window and do the resizing.

Is there any specific reason why we would want the references list outside of the main qf window?

**Comment by abaevbog** [comment]
> This sounds ideal both for the dragging behaviour and general reduced complexity.

I agree. @adomasven  does `-moz-window-dragging: drag` work for you as a replacement for the `WindowDraggingElement`? It works great on my mac but not on the windows VM, and I'm trying to track down what's wrong

**Comment by abaevbog** [comment]
Aha, so with this commit https://github.com/abaevbog/zotero/commit/c494169d1f2717268237510580e47a0c4af948c0, I have dragging kinda happening. I had to add `chromemargins` and remove 'hidechrome`. On mac it works fine but on windows, I have this:

https://github.com/zotero/zotero/assets/36271954/0fed6394-63b3-476c-aa82-706e062537e5

@windingwind, is that somewhat close to what you are seeing? Also, if I `return` right in the beginning of `_resizeWindow`, I still see this behavior. 

In any case, this feels relevant - whatever is going on there, it's 100% not supposed to be doing that so I'll investigate

**Comment by abaevbog** [comment]
Some findings I have to share:
- `chromemargins` is somehow what was causing the weird disappearance of the dialog in the video above. Initial value of '0,2,2,2' from `ZoteroPane` was causing such behavior but arbitrarily increasing it to '0,25,25,25' fixed the blinking/shrinking.
- I checked by essentially blocking all JS code and the effect from my video was still there without the `referencePanel`/window resizing/etc. Then it got fixed with `chromemargins` increased
- here I have the commit with the best working dragging so far with `-moz-window-drag` and no shrinking on my end: https://github.com/abaevbog/zotero/commit/f457ad14f237855d8c338b74a4aedfb4773c6612. There are a few issues: most importantly, the very top of the dialog doesn't seem to respond to dragging events, through the bottom or the sides do and follow the cursor without breaking away from it. Bubble dragging here is also overriden but as a proof of concept, I didn't worry about that yet.
- I cherry-picked this commit on top of the first 4 commits from https://github.com/zotero/zotero/pull/3889 (so `fx115` with the up-to-date QF without any extra fixes), and the dragging did work OK but the `panel` refused to follow the dialog just as it did before. So even if we do this, the panel dragging would need to be addressed. 

So yeah, I'll keep trying to iron this out but let me know if this all makes sense.

**Comment by windingwind** [comment]
it's weird to have such a large chromemargin. does it work with the '0,2,2,2,' if we do not use the margin on window?

**Comment by abaevbog** [comment]
It is weird and I am not sure why `chromemargin` has such effect. By just playing around with it, I found that `0,2,2,2` leads to the dialog [disappearing](https://github.com/zotero/zotero/issues/3847#issuecomment-2027565484) (at least on my win 11 vm), and increasing the bottom margin made it somewhat better. I finally settled on `0,0,15,0` - the bottom margin seems to be the one that matters.

**Comment by abaevbog** [comment]
>  does it work with the '0,2,2,2,' if we do not use the margin on window?

Sorry, which margin are you referring to? 

I did get a working setup without the old `windowDraggingUtils` by using `moz-window-drag` and working around some of its peculiarities. Most of it is in [this commit](https://github.com/zotero/zotero/pull/3889/commits/a90f833c89c2fc28b6205cc94170d91484050d06) from  https://github.com/zotero/zotero/pull/3889 and it works fairly well on all platforms that I could get my hands on. Let me know if you have any thoughts on what's going on there

**Comment by abaevbog** [comment]
Another report:  https://forums.zotero.org/discussion/113063/no-insert-citation-window-appears

**Comment by windingwind** [comment]
Thanks for all these efforts! I will try the commit and play with it a little. Ideally, being able to use one setting for non-titlebar windows would be great.

**Comment by adomasven** [comment]
Another report: https://forums.zotero.org/discussion/113360/zotero-7-beta-red-bar-disappears-but-open-documents-section-appears

**Comment by abaevbog** [comment]
Closing this now that we have a new redesigned citation dialog.
